File: AlgorithmLcr/Try1-Completely unknown single port fitting/cppversion/src/iofmt.cpp
#include "iofmt.hpp"

#include <cmath>
#include <cstdio>
#include <fstream>
#include <sstream>
#include <stdexcept>

namespace rlc {

namespace {

// Strip comments and blank lines (INPUT_FORMAT.md sec.0).
std::vector<std::string> contentLines(const std::string& text) {
    std::vector<std::string> out;
    std::istringstream is(text);
    std::string raw;
    while (std::getline(is, raw)) {
        size_t hash = raw.find('#');
        if (hash != std::string::npos) raw = raw.substr(0, hash);
        // trim
        size_t a = raw.find_first_not_of(" \t\r");
        size_t b = raw.find_last_not_of(" \t\r");
        if (a == std::string::npos) continue;
        out.push_back(raw.substr(a, b - a + 1));
    }
    return out;
}

[[noreturn]] void fail(const std::string& msg) {
    throw std::invalid_argument(msg);
}

std::string readFile(const std::string& path) {
    std::ifstream fh(path, std::ios::binary);
    if (!fh) fail("cannot open input file: " + path);
    std::ostringstream ss;
    ss << fh.rdbuf();
    return ss.str();
}

}  // namespace

Measurements parseMeasurements(const std::string& text) {
    auto rows = contentLines(text);
    if (rows.empty()) fail("empty measurement input");
    // first line: positive integer n
    size_t n = 0;
    try {
        n = (size_t)std::stoll(rows[0]);
    } catch (const std::exception&) {
        fail("first line must be integer n, got '" + rows[0] + "'");
    }
    if (n < 1) fail("n must be a positive integer");
    if (rows.size() - 1 != n)
        fail("n=" + std::to_string(n) + " but " +
             std::to_string(rows.size() - 1) + " data lines follow");

    Measurements out;
    out.f.reserve(n);
    out.z.reserve(n);
    for (size_t k = 0; k < n; ++k) {
        const std::string& line = rows[k + 1];
        std::istringstream ls(line);
        double fv = 0.0, rz = 0.0, iz = 0.0;
        std::string extra;
        if (!(ls >> fv >> rz >> iz)) fail("line " + std::to_string(k + 2) +
                                          ": expected 'f Rz Iz', got '" + line + "'");
        if (ls >> extra) fail("line " + std::to_string(k + 2) +
                              ": expected 3 fields, got '" + line + "'");
        if (!std::isfinite(fv) || !std::isfinite(rz) || !std::isfinite(iz))
            fail("line " + std::to_string(k + 2) + ": non-finite value");
        if (fv <= 0.0)
            fail("line " + std::to_string(k + 2) + ": frequency must be > 0");
        out.f.push_back(fv);
        out.z.emplace_back(rz, iz);
    }
    return out;
}

Measurements loadMeasurements(const std::string& path) {
    return parseMeasurements(readFile(path));
}

std::string formatMeasurements(const std::vector<double>& f,
                               const std::vector<std::complex<double>>& z) {
    if (f.size() != z.size())
        fail("f has " + std::to_string(f.size()) + " points, z has " +
             std::to_string(z.size()));
    if (f.empty()) fail("need at least one measurement point");
    std::ostringstream out;
    char buf[3][32];
    out << f.size() << "\n";
    for (size_t k = 0; k < f.size(); ++k) {
        if (!(std::isfinite(f[k]) && f[k] > 0.0))
            fail("frequency must be positive and finite");
        if (!std::isfinite(z[k].real()) || !std::isfinite(z[k].imag()))
            fail("impedance parts must be finite");
        std::snprintf(buf[0], sizeof(buf[0]), "%.17g", f[k]);
        std::snprintf(buf[1], sizeof(buf[1]), "%.17g", z[k].real());
        std::snprintf(buf[2], sizeof(buf[2]), "%.17g", z[k].imag());
        out << buf[0] << " " << buf[1] << " " << buf[2] << "\n";
    }
    return out.str();
}

int parseCount(const std::string& text) {
    auto rows = contentLines(text);
    if (rows.size() != 1)
        fail("count input must be exactly 1 line, got " +
             std::to_string(rows.size()));
    int n = 0;
    try {
        size_t pos = 0;
        long v = std::stol(rows[0], &pos);
        if (pos != rows[0].size()) throw std::invalid_argument("trailing");
        n = (int)v;
    } catch (const std::exception&) {
        fail("count line must be an integer, got '" + rows[0] + "'");
    }
    if (n < 1) fail("device count must be a positive integer");
    return n;
}

int loadCount(const std::string& path) { return parseCount(readFile(path)); }

std::string formatCount(int n) {
    if (n < 1) fail("device count must be a positive integer");
    return std::to_string(n) + "\n";
}

}  // namespace rlc

File: AlgorithmLcr/Try1-Completely unknown single port fitting/rlc_id/iofmt.py
"""Unified measurement input (../../INPUT_FORMAT.md section 1).

Text form: first line n, then exactly n lines ``f Rz Iz`` (frequency [Hz],
Re(Z) [ohm], Im(Z) [ohm]).  ``#`` starts a comment (whole line or trailing),
blank lines are ignored, fields are whitespace separated.  Dump uses %.17g
so load(dump(x)) == x bit-for-bit.
"""

from __future__ import annotations

import numpy as np


def _content_lines(text: str) -> list[str]:
    """Strip comments and blank lines (INPUT_FORMAT.md sec.0)."""
    out = []
    for raw in text.splitlines():
        s = raw.split("#", 1)[0].strip()
        if s:
            out.append(s)
    return out


def format_measurements(f, z) -> str:
    """Serialize (f, z) to the unified text form."""
    f = np.asarray(f, dtype=float)
    z = np.asarray(z, dtype=complex)
    if len(f) != len(z):
        raise ValueError(f"f has {len(f)} points, z has {len(z)}")
    if len(f) == 0:
        raise ValueError("need at least one measurement point")
    lines = [str(len(f))]
    for fk, zk in zip(f, z):
        if not (np.isfinite(fk) and fk > 0.0):
            raise ValueError(f"frequency must be positive and finite, got {fk}")
        if not (np.isfinite(zk.real) and np.isfinite(zk.imag)):
            raise ValueError(f"impedance parts must be finite, got {zk}")
        lines.append(f"{fk:.17g} {zk.real:.17g} {zk.imag:.17g}")
    return "\n".join(lines) + "\n"


def parse_measurements(text: str) -> tuple[np.ndarray, np.ndarray]:
    """Parse the unified text form -> (f, z).  Validation per sec.3."""
    rows = _content_lines(text)
    if not rows:
        raise ValueError("empty measurement input")
    try:
        n = int(rows[0])
    except ValueError:
        raise ValueError(f"first line must be integer n, got {rows[0]!r}") from None
    if n < 1:
        raise ValueError(f"n must be a positive integer, got {n}")
    data = rows[1:]
    if len(data) != n:
        raise ValueError(f"n={n} but {len(data)} data lines follow")
    fs, reals, imags = [], [], []
    for k, line in enumerate(data, start=2):  # 1-based line no. incl. n-line
        parts = line.split()
        if len(parts) != 3:
            raise ValueError(f"line {k}: expected 'f Rz Iz', got {line!r}")
        try:
            fv, rz, iz = (float(x) for x in parts)
        except ValueError:
            raise ValueError(f"line {k}: non-numeric field in {line!r}") from None
        if not np.isfinite([fv, rz, iz]).all():
            raise ValueError(f"line {k}: non-finite value in {line!r}")
        if fv <= 0.0:
            raise ValueError(f"line {k}: frequency must be > 0, got {fv}")
        fs.append(fv)
        reals.append(rz)
        imags.append(iz)
    return (np.asarray(fs, dtype=float),
            np.asarray(reals) + 1j * np.asarray(imags))


def load_measurements(path) -> tuple[np.ndarray, np.ndarray]:
    """Read the unified measurement file from disk."""
    with open(path, "r", encoding="utf-8") as fh:
        return parse_measurements(fh.read())
